 ## Importing Libraries

In [316]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, GridSearchCV


import warnings
warnings.filterwarnings('ignore')

## Importing Dataset

In [317]:
df = pd.read_csv('Terry_Stops.csv')

df = df_copy

df.head()

Subject Age Group  Subject ID     GO / SC Num  Terry Stop ID  \
0                 -          -1  20140000120677          92317   
1                 -          -1  20150000001463          28806   
2                 -          -1  20150000001516          29599   
3                 -          -1  20150000001670          32260   
4                 -          -1  20150000001739          33155   

  Stop Resolution Weapon Type Officer ID  Officer YOB Officer Gender  \
0          Arrest        None     7500           1984              M   
1   Field Contact        None     5670           1965              M   
2   Field Contact        None     4844           1961              M   
3   Field Contact        None     7539           1963              M   
4   Field Contact        None     6973           1977              M   

                Officer Race     Subject Perceived Race  \
0  Black or African American                      Asian   
1                      White                          -   
2                      White                      White   
3                      White                          -   
4                      White  Black or African American   

  Subject Perceived Gender        Reported Date Reported Time  \
0                     Male  2015-10-16T00:00:00      11:32:00   
1                        -  2015-03-19T00:00:00      07:59:00   
2                     Male  2015-03-21T00:00:00      19:12:00   
3                        -  2015-04-01T00:00:00      04:55:00   
4                     Male  2015-04-03T00:00:00      00:41:00   

  Initial Call Type Final Call Type Call Type             Officer Squad  \
0                 -               -         -  SOUTH PCT 1ST W - ROBERT   
1                 -               -         -                       NaN   
2                 -               -         -                       NaN   
3                 -               -         -                       NaN   
4                 -               -         -                       NaN   

  Arrest Flag Frisk Flag Precinct  Sector    Beat  
0           N          N    South  O       O2      
1           N          N        -       -       -  
2           N          -        -       -       -  
3           N          N        -       -       -  
4           N          N        -       -       -

In [318]:
# checking for null, data type and name of columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44838 entries, 0 to 44837
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Subject Age Group         44838 non-null  object
 1   Subject ID                44838 non-null  int64 
 2   GO / SC Num               44838 non-null  int64 
 3   Terry Stop ID             44838 non-null  int64 
 4   Stop Resolution           44838 non-null  object
 5   Weapon Type               44838 non-null  object
 6   Officer ID                44838 non-null  object
 7   Officer YOB               44838 non-null  int64 
 8   Officer Gender            44838 non-null  object
 9   Officer Race              44838 non-null  object
 10  Subject Perceived Race    44838 non-null  object
 11  Subject Perceived Gender  44838 non-null  object
 12  Reported Date             44838 non-null  object
 13  Reported Time             44838 non-null  object
 14  Initial Call Type     

In [319]:
# checking for null values
df.isna().sum()

Subject Age Group             0
Subject ID                    0
GO / SC Num                   0
Terry Stop ID                 0
Stop Resolution               0
Weapon Type                   0
Officer ID                    0
Officer YOB                   0
Officer Gender                0
Officer Race                  0
Subject Perceived Race        0
Subject Perceived Gender      0
Reported Date                 0
Reported Time                 0
Initial Call Type             0
Final Call Type               0
Call Type                     0
Officer Squad               580
Arrest Flag                   0
Frisk Flag                    0
Precinct                      0
Sector                        0
Beat                          0
dtype: int64

In [320]:
# drop the rows with null values and reset the index
df = df.dropna()
df = df.reset_index(drop=True)
df.isna().sum()

Subject Age Group           0
Subject ID                  0
GO / SC Num                 0
Terry Stop ID               0
Stop Resolution             0
Weapon Type                 0
Officer ID                  0
Officer YOB                 0
Officer Gender              0
Officer Race                0
Subject Perceived Race      0
Subject Perceived Gender    0
Reported Date               0
Reported Time               0
Initial Call Type           0
Final Call Type             0
Call Type                   0
Officer Squad               0
Arrest Flag                 0
Frisk Flag                  0
Precinct                    0
Sector                      0
Beat                        0
dtype: int64

In [321]:
# check for duplicates
boolean = df.duplicated().any()
print(boolean, end='\n\n') # True

False



In [322]:
df = df.replace('-', 'Unknown')

In [323]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44258 entries, 0 to 44257
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Subject Age Group         44258 non-null  object
 1   Subject ID                44258 non-null  int64 
 2   GO / SC Num               44258 non-null  int64 
 3   Terry Stop ID             44258 non-null  int64 
 4   Stop Resolution           44258 non-null  object
 5   Weapon Type               44258 non-null  object
 6   Officer ID                44258 non-null  object
 7   Officer YOB               44258 non-null  int64 
 8   Officer Gender            44258 non-null  object
 9   Officer Race              44258 non-null  object
 10  Subject Perceived Race    44258 non-null  object
 11  Subject Perceived Gender  44258 non-null  object
 12  Reported Date             44258 non-null  object
 13  Reported Time             44258 non-null  object
 14  Initial Call Type     

In [324]:
#look at the unique data of each column
for col in df.columns:
    print(col, '\n', df[col].unique(), '\n\n')

Subject Age Group 
 ['Unknown' '46 - 55' '56 and Above' '1 - 17' '18 - 25' '36 - 45' '26 - 35'] 


Subject ID 
 [         -1 16366865981  7726713382 ... 13855696527 13930427243
 14268071835] 


GO / SC Num 
 [20140000120677 20150000002351 20150000002363 ... 20200000211163
 20200000218545 20200000223475] 


Terry Stop ID 
 [      92317       45252       45182 ... 13855817033 13928574375
 14273492771] 


Stop Resolution 
 ['Arrest' 'Field Contact' 'Citation / Infraction' 'Offense Report'
 'Referred for Prosecution'] 


Weapon Type 
 ['None' 'Firearm Other' 'Lethal Cutting Instrument' 'Handgun' 'Unknown'
 'Club, Blackjack, Brass Knuckles' 'Knife/Cutting/Stabbing Instrument'
 'Other Firearm' 'Rifle' 'Fire/Incendiary Device'
 'Blunt Object/Striking Implement' 'Firearm (unk type)'
 'Mace/Pepper Spray' 'None/Not Applicable' 'Club' 'Firearm'
 'Taser/Stun Gun' 'Shotgun' 'Brass Knuckles' 'Automatic Handgun'
 'Blackjack'] 


Officer ID 
 ['7500  ' '7591  ' '7560  ' ... '8749  ' '6983  ' '5917  ']

In [325]:
df.describe(include = 'all')

Subject Age Group    Subject ID   GO / SC Num  Terry Stop ID  \
count              44258  4.425800e+04  4.425800e+04   4.425800e+04   
unique                 7           NaN           NaN            NaN   
top              26 - 35           NaN           NaN            NaN   
freq               14743           NaN           NaN            NaN   
mean                 NaN  2.061739e+09  2.017454e+13   2.763394e+09   
std                  NaN  3.937583e+09  1.621511e+10   5.149534e+09   
min                  NaN -1.000000e+00  2.009000e+13   4.257800e+04   
25%                  NaN -1.000000e+00  2.016000e+13   1.772252e+05   
50%                  NaN -1.000000e+00  2.018000e+13   3.880850e+05   
75%                  NaN -1.000000e+00  2.019000e+13   5.512335e+05   
max                  NaN  1.755179e+10  2.020000e+13   1.755178e+10   

       Stop Resolution Weapon Type Officer ID   Officer YOB Officer Gender  \
count            44258       44258      44258  44258.000000          44258   
unique               5          21       1164           NaN              2   
top      Field Contact        None     7456             NaN              M   
freq             17704       32085        393           NaN          39226   
mean               NaN         NaN        NaN   1982.779882            NaN   
std                NaN         NaN        NaN      9.004519            NaN   
min                NaN         NaN        NaN   1900.000000            NaN   
25%                NaN         NaN        NaN   1978.000000            NaN   
50%                NaN         NaN        NaN   1985.000000            NaN   
75%                NaN         NaN        NaN   1989.000000            NaN   
max                NaN         NaN        NaN   1997.000000            NaN   

       Officer Race Subject Perceived Race Subject Perceived Gender  \
count         44258                  44258                    44258   
unique            9                      9                        5   
top           White                  White                     Male   
freq          33689                  21631                    34593   
mean            NaN                    NaN                      NaN   
std             NaN                    NaN                      NaN   
min             NaN                    NaN                      NaN   
25%             NaN                    NaN                      NaN   
50%             NaN                    NaN                      NaN   
75%             NaN                    NaN                      NaN   
max             NaN                    NaN                      NaN   

              Reported Date Reported Time Initial Call Type Final Call Type  \
count                 44258         44258             44258           44258   
unique                 1998         10894               165             204   
top     2015-10-01T00:00:00      17:00:00           Unknown         Unknown   
freq                    101            51             12791           12791   
mean                    NaN           NaN               NaN             NaN   
std                     NaN           NaN               NaN             NaN   
min                     NaN           NaN               NaN             NaN   
25%                     NaN           NaN               NaN             NaN   
50%                     NaN           NaN               NaN             NaN   
75%                     NaN           NaN               NaN             NaN   
max                     NaN           NaN               NaN             NaN   

       Call Type                    Officer Squad Arrest Flag Frisk Flag  \
count      44258                            44258       44258      44258   
unique         7                              168           2          3   
top          911  TRAINING - FIELD TRAINING SQUAD           N          N   
freq       19668                             4756       41633      33966   
mean         NaN                   

In [326]:
#look at the value counts of each column (not normalized)
def col_values(df):
    """
    For use in Preprocessing and cleaning to find placeholder values
    Input: Data frame
    Output: Counts of unique values for each column
    """
    for col in df.columns:
        print(col)
        print('\n')
        print(df[col].value_counts())
        print('-------------------------------------------------------')
        print('\n')
        
col_values(df)

Subject Age Group


26 - 35         14743
36 - 45          9335
18 - 25          8924
46 - 55          5698
56 and Above     2251
1 - 17           1907
Unknown          1400
Name: Subject Age Group, dtype: int64
-------------------------------------------------------


Subject ID


-1              34181
 7726859935        19
 7753260438        13
 7727117712        12
 7734218049         8
                ...  
 7730602336         1
 15879606630        1
 7758922092         1
 7727677812         1
 16219707395        1
Name: Subject ID, Length: 7899, dtype: int64
-------------------------------------------------------


GO / SC Num


20160000378750    16
20150000190790    16
20180000134604    14
20190000441736    13
20170000132836    13
                  ..
20190000045451     1
20170000158799     1
20150000003136     1
20180000479302     1
20180000071981     1
Name: GO / SC Num, Length: 34574, dtype: int64
-------------------------------------------------------


Terry Stop ID


130800

Initial Thoughts on How to Handle the Columns:

**Subject Age Group** none
**Subject ID**  drop since subjects are not required to present ID
**GO / SC Num**  drop because the numbers may have a one to many relationship
**Terry Stop ID**  drop as ID is not important when looking at relationships
**Stop Resolution** Categorize to Arrest(1) and Not Arrested(0)
**Weapon Type** Categorize to Weapon found at time of Frisk Yes(1) No(0)
**Officer ID** Create Bins of 0-99, 100-199, 200-299, 300-399
    Q: Were Officers with higher amount of cases/stops having high arrest percentage?
**Officer YOB**  Bin the birthdates
**Officer Gender**  Categorize to Female(1), Male(0)
**Officer Race**  Categorize to White(1), Not-White(0)
**Subject Perceived Race**  Drop as this is based on the officers' account and can be biased
**Subject Perceived Gender**  Categorize to only Female(1) and Male(0)
**Reported Date**  Categorize to Years (maybe bins)
**Reported Time**  Categorize to Day and Night shifts
**Initial Call Type**  Drop since majority is categorized as unknown
**Final Call Type** Drop since majority is categorized as unknown
**Call Type**  Categorize as 911(1) and not911(0)
**Officer Squad**  Drop too much room for error in trying to evaluate this data set
**Arrest Flag** Good
**Frisk Flag**  Drop rows with 'unknown' as this is a small amount
**Precinct** fix OOJ and FK ERROR. May help with location of dispatched officer
**Sector**. May help with location of dispatched officer
Beat**  Precinct,Sector and Beat are together with identifying location of dispatch

In [327]:
# show names of columns
for col in df.columns:
    print(col)

Subject Age Group
Subject ID
GO / SC Num
Terry Stop ID
Stop Resolution
Weapon Type
Officer ID
Officer YOB
Officer Gender
Officer Race
Subject Perceived Race
Subject Perceived Gender
Reported Date
Reported Time
Initial Call Type
Final Call Type
Call Type
Officer Squad
Arrest Flag
Frisk Flag
Precinct
Sector
Beat


Take out the columns I don't need and make the columns python friendly 

In [328]:
df.columns = ['subject_age_group', 'subject_id', 'go_sc_num', 'terry_stop_id',
       'stop_resolution', 'weapon_type', 'officer_id', 'officer_yob',
       'officer_gender', 'officer_race', 'subject_perceived_race',
       'subject_perceived_gender', 'reported_date', 'reported_time',
       'initial_call_type', 'final_call_type', 'call_type', 'officer_squad',
       'arrest_flag', 'frisk_flag', 'precinct', 'sector', 'beat']
df.columns


Index(['subject_age_group', 'subject_id', 'go_sc_num', 'terry_stop_id',
       'stop_resolution', 'weapon_type', 'officer_id', 'officer_yob',
       'officer_gender', 'officer_race', 'subject_perceived_race',
       'subject_perceived_gender', 'reported_date', 'reported_time',
       'initial_call_type', 'final_call_type', 'call_type', 'officer_squad',
       'arrest_flag', 'frisk_flag', 'precinct', 'sector', 'beat'],
      dtype='object')

## **Scrubbing**

In [329]:
df.head()

subject_age_group  subject_id       go_sc_num  terry_stop_id  \
0           Unknown          -1  20140000120677          92317   
1           Unknown          -1  20150000002351          45252   
2           Unknown          -1  20150000002363          45182   
3           Unknown          -1  20150000002392          45365   
4           Unknown          -1  20150000002451          46430   

  stop_resolution weapon_type officer_id  officer_yob officer_gender  \
0          Arrest        None     7500           1984              M   
1   Field Contact        None     7591           1985              M   
2   Field Contact        None     7591           1985              M   
3   Field Contact        None     7560           1986              M   
4   Field Contact        None     7591           1985              M   

                officer_race subject_perceived_race subject_perceived_gender  \
0  Black or African American                  Asian                     Male   
1         Hispanic or Latino                  Other                   Female   
2         Hispanic or Latino                  White                     Male   
3                      White                  White                   Female   
4         Hispanic or Latino                Unknown                  Unknown   

         reported_date reported_time initial_call_type final_call_type  \
0  2015-10-16T00:00:00      11:32:00           Unknown         Unknown   
1  2015-05-20T00:00:00      21:39:00           Unknown         Unknown   
2  2015-05-20T00:00:00      22:40:00           Unknown         Unknown   
3  2015-05-22T00:00:00      07:39:00           Unknown         Unknown   
4  2015-05-25T00:00:00      01:06:00           Unknown         Unknown   

  call_type                 officer_squad arrest_flag frisk_flag precinct  \
0   Unknown      SOUTH PCT 1ST W - ROBERT           N          N    South   
1   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   
2   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   
3   Unknown  SOUTH PCT 1ST W - R/S RELIEF           N          N     East   
4   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   

    sector     beat  
0   O        O2      
1  Unknown  Unknown  
2  Unknown  Unknown  
3   E        E2      
4  Unknown  Unknown

In [330]:
# df = df.drop(['go_sc_num','terry_stop_id','officer_id','sector','beat'], axis=1)
# df.columns

### Subject ID

Will need to feature engineer this to make a column for repeat offenders and drop subject_id

In [331]:
import numpy as np

df['repeat_offenders'] = np.nan

for i in range(len(df['subject_id'])):
    if df['subject_id'].iloc[i] > 1:
        df['repeat_offenders'].iloc[i] = 'Y'
    else:
        df['repeat_offenders'].iloc[i] = 'N'

df['repeat_offenders'].value_counts()

N    34181
Y    10077
Name: repeat_offenders, dtype: int64

In [332]:
df = df.drop(['subject_id'], axis=1)

In [333]:
df.head()

subject_age_group       go_sc_num  terry_stop_id stop_resolution  \
0           Unknown  20140000120677          92317          Arrest   
1           Unknown  20150000002351          45252   Field Contact   
2           Unknown  20150000002363          45182   Field Contact   
3           Unknown  20150000002392          45365   Field Contact   
4           Unknown  20150000002451          46430   Field Contact   

  weapon_type officer_id  officer_yob officer_gender  \
0        None     7500           1984              M   
1        None     7591           1985              M   
2        None     7591           1985              M   
3        None     7560           1986              M   
4        None     7591           1985              M   

                officer_race subject_perceived_race subject_perceived_gender  \
0  Black or African American                  Asian                     Male   
1         Hispanic or Latino                  Other                   Female   
2         Hispanic or Latino                  White                     Male   
3                      White                  White                   Female   
4         Hispanic or Latino                Unknown                  Unknown   

         reported_date reported_time initial_call_type final_call_type  \
0  2015-10-16T00:00:00      11:32:00           Unknown         Unknown   
1  2015-05-20T00:00:00      21:39:00           Unknown         Unknown   
2  2015-05-20T00:00:00      22:40:00           Unknown         Unknown   
3  2015-05-22T00:00:00      07:39:00           Unknown         Unknown   
4  2015-05-25T00:00:00      01:06:00           Unknown         Unknown   

  call_type                 officer_squad arrest_flag frisk_flag precinct  \
0   Unknown      SOUTH PCT 1ST W - ROBERT           N          N    South   
1   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   
2   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   
3   Unknown  SOUTH PCT 1ST W - R/S RELIEF           N          N     East   
4   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   

    sector     beat repeat_offenders  
0   O        O2                    N  
1  Unknown  Unknown                N  
2  Unknown  Unknown                N  
3   E        E2                    N  
4  Unknown  Unknown                N

In [334]:
df.weapon_type.value_counts()

None                                 32085
Unknown                               9607
Lethal Cutting Instrument             1454
Knife/Cutting/Stabbing Instrument      494
Handgun                                278
Firearm Other                           98
Blunt Object/Striking Implement         66
Club, Blackjack, Brass Knuckles         49
Firearm                                 34
Mace/Pepper Spray                       20
Other Firearm                           18
Firearm (unk type)                      15
Club                                     8
Rifle                                    7
Taser/Stun Gun                           7
None/Not Applicable                      7
Fire/Incendiary Device                   4
Shotgun                                  3
Automatic Handgun                        2
Blackjack                                1
Brass Knuckles                           1
Name: weapon_type, dtype: int64

In [335]:
df['weapon_type']=df['weapon_type'].replace('-', 'Type Unknown')

In [336]:
df.weapon_type.value_counts()

None                                 32085
Unknown                               9607
Lethal Cutting Instrument             1454
Knife/Cutting/Stabbing Instrument      494
Handgun                                278
Firearm Other                           98
Blunt Object/Striking Implement         66
Club, Blackjack, Brass Knuckles         49
Firearm                                 34
Mace/Pepper Spray                       20
Other Firearm                           18
Firearm (unk type)                      15
Club                                     8
Rifle                                    7
Taser/Stun Gun                           7
None/Not Applicable                      7
Fire/Incendiary Device                   4
Shotgun                                  3
Automatic Handgun                        2
Blackjack                                1
Brass Knuckles                           1
Name: weapon_type, dtype: int64

In [337]:
# Switch Weapon Type to Weapon (0's & 1's)
Weapon = [] # Initialize a an Empty List
for value in df['weapon_type']:
    if value == 'None': # Replace "None" Values with "0" 
        Weapon.append(0)
    else:
        Weapon.append(1) # Replace Other Values with "0" 
#df['weapon_type'] = Weapon # Create a New Column Called "Weapon"
pd.options.display.max_columns = None  # Print First 5 Rows of DataFrame
display(df)

subject_age_group       go_sc_num  terry_stop_id stop_resolution  \
0               Unknown  20140000120677          92317          Arrest   
1               Unknown  20150000002351          45252   Field Contact   
2               Unknown  20150000002363          45182   Field Contact   
3               Unknown  20150000002392          45365   Field Contact   
4               Unknown  20150000002451          46430   Field Contact   
...                 ...             ...            ...             ...   
44253      56 and Above  20200000209169    13852344465   Field Contact   
44254      56 and Above  20200000210914    13855407067  Offense Report   
44255      56 and Above  20200000211163    13855817033          Arrest   
44256      56 and Above  20200000218545    13928574375   Field Contact   
44257      56 and Above  20200000223475    14273492771          Arrest   

                             weapon_type officer_id  officer_yob  \
0                                   None     7500           1984   
1                                   None     7591           1985   
2                                   None     7591           1985   
3                                   None     7560           1986   
4                                   None     7591           1985   
...                                  ...        ...          ...   
44253  Knife/Cutting/Stabbing Instrument     8537           1988   
44254                            Unknown     8465           1990   
44255                            Unknown     8626           1992   
44256                            Unknown     8672           1995   
44257                            Firearm     8731           1994   

      officer_gender               officer_race     subject_perceived_race  \
0                  M  Black or African American                      Asian   
1                  M         Hispanic or Latino                      Other   
2                  M         Hispanic or Latino                      White   
3                  M                      White                      White   
4                  M         Hispanic or Latino                    Unknown   
...              ...                        ...                        ...   
44253              M  Black or African American  Black or African American   
44254              M                      White                      White   
44255              M                      White                      White   
44256              F         Hispanic or Latino                      White   
44257              M                      Asian  Black or African American   

      subject_perceived_gender        reported_date reported_time  \
0                         Male  2015-10-16T00:00:00      11:32:00   
1                       Female  2015-05-20T00:00:00      21:39:00   
2                         Male  2015-05-20T00:00:00      22:40:00   
3                       Female  2015-05-22T00:00:00      07:39:00   
4                      Unknown  2015-05-25T00:00:00      01:06:00   
...                        ...                  ...           ...   
44253                     Male  2020-07-09T00:00:00      19:11:09   
44254                     Male  2020-07-11T00:00:00      23:03:10   
44255                     Male  2020-07-12T00:00:00      08:20:33   
44256                     Male  2020-07-21T00:00:00      00:41:13   
44257                     Male  2020-07-26T00:00:00      20:37:41   

                                    initial_call_type  \
0                                             Unknown   
1                                             Unknown   
2                                             Unknown   
3                                             Unknown   
4                                             Unknown   
...                                               ...   
44253  ASLT - IP/JO - WITH OR W/O WPNS (NO SHOOTINGS)   
44254   THREATS (INCLS IN-PERSON/BY PHONE/IN WRITING)   
44255                

No Weapons Found = 0, Weapons Found = 1

In [338]:
df.weapon_type.value_counts()

None                                 32085
Unknown                               9607
Lethal Cutting Instrument             1454
Knife/Cutting/Stabbing Instrument      494
Handgun                                278
Firearm Other                           98
Blunt Object/Striking Implement         66
Club, Blackjack, Brass Knuckles         49
Firearm                                 34
Mace/Pepper Spray                       20
Other Firearm                           18
Firearm (unk type)                      15
Club                                     8
Rifle                                    7
Taser/Stun Gun                           7
None/Not Applicable                      7
Fire/Incendiary Device                   4
Shotgun                                  3
Automatic Handgun                        2
Blackjack                                1
Brass Knuckles                           1
Name: weapon_type, dtype: int64

In [339]:
df.shape

(44258, 23)

In [340]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44258 entries, 0 to 44257
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   subject_age_group         44258 non-null  object
 1   go_sc_num                 44258 non-null  int64 
 2   terry_stop_id             44258 non-null  int64 
 3   stop_resolution           44258 non-null  object
 4   weapon_type               44258 non-null  object
 5   officer_id                44258 non-null  object
 6   officer_yob               44258 non-null  int64 
 7   officer_gender            44258 non-null  object
 8   officer_race              44258 non-null  object
 9   subject_perceived_race    44258 non-null  object
 10  subject_perceived_gender  44258 non-null  object
 11  reported_date             44258 non-null  object
 12  reported_time             44258 non-null  object
 13  initial_call_type         44258 non-null  object
 14  final_call_type       

In [341]:
df.call_type.value_counts()

911                              19668
Unknown                          12791
ONVIEW                            8422
TELEPHONE OTHER, NOT 911          3079
ALARM CALL (NOT POLICE ALARM)      291
TEXT MESSAGE                         6
SCHEDULED EVENT (RECURRING)          1
Name: call_type, dtype: int64

In [342]:
# Switch Call Type to Call Made/No Call Made (0's & 1's)
call_made = [] # Initialize a an Empty List
for value in df['call_type']:
    if value == '-':  
        call_made.append(0) # Replace "-" Values with "0"
    else:
        call_made.append(1) # Replace Other Values with "1"

df.head() # Print First 5 Rows of DataFrame


subject_age_group       go_sc_num  terry_stop_id stop_resolution  \
0           Unknown  20140000120677          92317          Arrest   
1           Unknown  20150000002351          45252   Field Contact   
2           Unknown  20150000002363          45182   Field Contact   
3           Unknown  20150000002392          45365   Field Contact   
4           Unknown  20150000002451          46430   Field Contact   

  weapon_type officer_id  officer_yob officer_gender  \
0        None     7500           1984              M   
1        None     7591           1985              M   
2        None     7591           1985              M   
3        None     7560           1986              M   
4        None     7591           1985              M   

                officer_race subject_perceived_race subject_perceived_gender  \
0  Black or African American                  Asian                     Male   
1         Hispanic or Latino                  Other                   Female   
2         Hispanic or Latino                  White                     Male   
3                      White                  White                   Female   
4         Hispanic or Latino                Unknown                  Unknown   

         reported_date reported_time initial_call_type final_call_type  \
0  2015-10-16T00:00:00      11:32:00           Unknown         Unknown   
1  2015-05-20T00:00:00      21:39:00           Unknown         Unknown   
2  2015-05-20T00:00:00      22:40:00           Unknown         Unknown   
3  2015-05-22T00:00:00      07:39:00           Unknown         Unknown   
4  2015-05-25T00:00:00      01:06:00           Unknown         Unknown   

  call_type                 officer_squad arrest_flag frisk_flag precinct  \
0   Unknown      SOUTH PCT 1ST W - ROBERT           N          N    South   
1   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   
2   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   
3   Unknown  SOUTH PCT 1ST W - R/S RELIEF           N          N     East   
4   Unknown         WEST PCT 3RD W - MARY           N          N  Unknown   

    sector     beat repeat_offenders  
0   O        O2                    N  
1  Unknown  Unknown                N  
2  Unknown  Unknown                N  
3   E        E2                    N  
4  Unknown  Unknown                N

In [343]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44258 entries, 0 to 44257
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   subject_age_group         44258 non-null  object
 1   go_sc_num                 44258 non-null  int64 
 2   terry_stop_id             44258 non-null  int64 
 3   stop_resolution           44258 non-null  object
 4   weapon_type               44258 non-null  object
 5   officer_id                44258 non-null  object
 6   officer_yob               44258 non-null  int64 
 7   officer_gender            44258 non-null  object
 8   officer_race              44258 non-null  object
 9   subject_perceived_race    44258 non-null  object
 10  subject_perceived_gender  44258 non-null  object
 11  reported_date             44258 non-null  object
 12  reported_time             44258 non-null  object
 13  initial_call_type         44258 non-null  object
 14  final_call_type       

In [344]:
df.officer_gender.value_counts()

M    39226
F     5032
Name: officer_gender, dtype: int64

In [345]:
# Switch Call Type to Call Made/No Call Made (0's & 1's)
Gender_of_Officer = [] # Initialize a an Empty List
for value in df['officer_gender']:
    if value == 'M':  
        call_made.append(0) # Replace "-" Values with "0"
    else:
        call_made.append(1) # Replace Other Values with "1"
df['officer_gender'] = Gender_of_Officer # Create a New Column Called "Call Made"
df = df.drop(['officer_gender'], axis=1)
df.head() # Print First 5 Rows of DataFrame

ValueError: Length of values does not match length of index